server/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java

```diff
@@ -204,6 +217,115 @@ public boolean shouldCancelChildrenOnCancellation() {
+
+        public static Set<String> FIELDS_SET = new HashSet<>();
+        static {
+            FIELDS_SET.add(SLICE_ID_FIELD);
+            FIELDS_SET.add(TOTAL_FIELD);
+            FIELDS_SET.add(UPDATED_FIELD);
+            FIELDS_SET.add(CREATED_FIELD);
+            FIELDS_SET.add(DELETED_FIELD);
+            FIELDS_SET.add(BATCHES_FIELD);
+            FIELDS_SET.add(VERSION_CONFLICTS_FIELD);
+            FIELDS_SET.add(NOOPS_FIELD);
+            FIELDS_SET.add(RETRIES_FIELD);
+            // No need for inner level fields for retries in the set of outer level fields
+            FIELDS_SET.add(THROTTLED_RAW_FIELD);
+            FIELDS_SET.add(THROTTLED_HR_FIELD);
+            FIELDS_SET.add(REQUESTS_PER_SEC_FIELD);
+            FIELDS_SET.add(CANCELED_FIELD);
+            FIELDS_SET.add(THROTTLED_UNTIL_RAW_FIELD);
+            FIELDS_SET.add(THROTTLED_UNTIL_HR_FIELD);
+            FIELDS_SET.add(SLICES_FIELD);
+        }
+
+        @SuppressWarnings("unchecked")
+        static ConstructingObjectParser<Tuple<Long, Long>, Void> RETRIES_PARSER = new ConstructingObjectParser<>(
+            "bulk_by_scroll_task_status_retries",
+            true,
+            a -> new Tuple(a[0], a[1])
+        );
+        static {
+            RETRIES_PARSER.declareLong(constructorArg(), new ParseField(RETRIES_BULK_FIELD));
+            RETRIES_PARSER.declareLong(constructorArg(), new ParseField(RETRIES_SEARCH_FIELD));
+        }
+
+        /**
+         * Tries to construct the object based on the order on which fields were declared
+         * in {@link #declareFields}
+         * @param a the array of objects returned by ObjectParser
+         * @param startIndex the startIndex from which to start reading in this array
+         * @return the constructed Status object
+         */
+        @SuppressWarnings("unchecked")
+        public static Status constructFromObjectArray(Object[] a, int startIndex) {
+            Integer sliceId = (Integer) a[startIndex];
+            Long total = (Long) a[startIndex + 1];
+            Long updated = (Long) a[startIndex + 2];
+            Long created = (Long) a[startIndex + 3];
+            Long deleted = (Long) a[startIndex + 4];
+            Integer batches = (Integer) a[startIndex + 5];
+            Long versionConflicts = (Long) a[startIndex + 6];
+            Long noops = (Long) a[startIndex + 7];
+            Tuple<Long, Long> retries = (Tuple<Long, Long>) a[startIndex + 8];
+            Long bulkRetries = retries.v1();
+            Long searchRetries = retries.v2();
+            TimeValue throttled =
+                a[startIndex + 9] != null ? new TimeValue((Long)a[startIndex + 9], TimeUnit.MILLISECONDS) : null;
+            Float requestsPerSecond = (Float) a[startIndex + 10];
+            requestsPerSecond = requestsPerSecond == -1 ? Float.POSITIVE_INFINITY : requestsPerSecond;
+            String reasonCancelled = (String) a[startIndex + 11];
+            TimeValue throttledUntil =
```

Comment:
I'd probably extract them all from the array first, casting them, and then manipulating them.


Actually, I wonder if a builder object would be better than this object array stuff. It might be cleaner. Not nice, but the way I laid out the json makes nice kind of impossible.

Reply:
I see what you mean. But it might not be super helpful. There are two constructors. The way it works right now is you call one or the other. So you calculate stuff from `sliceStatuses` or you specify everything and `sliceStatuses` is an empty list. Not sure how that would translate here. Also, personally I like immutable objects. Validation is consolidated in the constructor. For example, fields that must be set. Plus it wouldn't change much in terms of lines of code I think.
But if you think it's important I can change it.

Reply:
By mutable builder I mean making a mutable object that you just use as the target for the parsing logic and then throw away afterwords. It just has setters and then a method to build status or error depending on which fields are set. The mutable object never escapes parsing. It might not be less code but I *think* it'd be easier to understand.